### PR TITLE
Cadence 5.1

### PR DESCRIPTION
**Pull request by kenellorando**
# Cadence Radio 5.1 Release Notes

## Development Mode
A new development mode is available that may be enabled by setting Cadence's environment `CSERVER_DEVMODE` to `1`. Devmode exposes reserved API routes to perform administrative actions. I find this useful for when I need to speed up certain actions to manually test things, like skip to next song (see Skip Track API).

It goes without saying, but don't enable devmode on a production server.

## Get History for Last Ten Played Songs
5.1 introduces a song history API which returns the last ten played tracks' titles, artists, and ending-times. This is useful for listeners that heard a song but did not get its name. The route can be found at `GET /api/history`.

The UI is updated with a new tab which demonstrates history API usage. The UI calculates when a song ended in relative terms, like "just now", or "15 minutes ago". 

A new server-sent event type called `history` is also introduced, which sends no relevant data but can be used as a useful signal that a song has changed. This event type is the UI's new method for knowing when to update now playing information and history.

## Get Icecast Bitrate
An endpoint for getting stream bitrate is also available under `GET /api/bitrate`. If a stream is actively playing under Cadence Icecast, it will return the integer value of the stream's bitrate in kilobits. For example, a return of `192` means the stream is running at `192kbps`.

## Skip the Currently Playing Track
A devmode route is available under `GET /api/dev/skip`. As the name implies, calling this API skips the currently playing track.

## Passive Music Directory Watch and Database (Re)Population
Cadence will now automatically watch the provided `CSERVER_MUSICDIR` path (non-recursively) for any file changes. If you remove a music file or add one to the directory while the Cadence API server is actively running, Cadence will repopulate the database automatically. In other words, you no longer need to restart the server when you make music changes.

## UI "Artwork" Support for Tracks with no Artwork
The UI has been updated to display a default "empty" album art photo when a song with no album artwork is selected for play.

# Changed, Deferred, Cancelled
| Status | Update                                | Comments                                                                                                                                                                                                         |
|--------|---------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ✔️      | Optional  _Development Mode_          |                                                                                                                                                                                                                  |
| ✔️      | History API                           |                                                                                                                                                                                                                  |
| ✔️      | Bitrate API                           |                                                                                                                                                                                                                  |
| ✔️      | Skip Track API                        |                                                                                                                                                                                                                  |
| ✔️      | Passive Music Directory Watch         |                                                                                                                                                                                                                  |
| ✔️      | Database Auto-Repopulation            | It won't update the frontend right away though, so that's something we'll have to keep an eye on.                                                                                                                |
| ✔️      | 'Default' Album Artwork               |                                                                                                                                                                                                                  |
| ✔️      | Updates to Base Images for Components |                                                                                                                                                                                                                  |
| ✔️      | Styling Updates                       |                                                                                                                                                                                                                  |
| ✔️      | Light Code Cleanup                    | Less overall global variables. More usage of structs. Less redundancies.                                                                                                                                         |
| ⏩      | Safer Configuration File Handling     | Ended up not making this release. I am still considering alternatives to config handling.                                                                                                                        |
| ❌      | Rebuild the Theme Changer             | Cancelled. It's simply too difficult to theme over a CSS library using plain CSS. I'm instead looking at the possibility of just completely overhauling the entire frontend using a modern JavaScript framework. |

### CODE REVIEW

**Comment by za419** [comment]
Oh boy, new features to add to za419/CadenceBot once I get around to fixing it in general!

... And decide whether it should broadcast at the same bitrate as Cadence, or obey config/maximum rate anyway to preserve data through compression.

Agh! Good stuff, though, @kenellorando!

File: cadence/server/api_actions.go [hunk 1]
```diff
@@ -121,85 +130,136 @@ func liquidsoapRequest(path string) (message string, err error) {
...
+	fmt.Fprintf(conn, "quit"+"\n")
+	return message, nil
+}
 
-// Takes no arguments.
-// Returns nothing, but sends updated stream info to SSE and sets the global variables.
-// It is launched as a goroutine by init.
+// Watches the music directory (CSERVER_MUSICDIR) for any changes, and reconfigures the database.
+func filesystemMonitor() {
+	watcher, err := fsnotify.NewWatcher()
```
Comment (za419) [inline cadence/server/api_actions.go:150]: Are we considering updating to a recursive method (rfsnotify/roll our own/etc), or are we going to constrain that if you want updates to appear you need to make sure it touches the mtime on the MusicDir?

File: cadence/server/api_actions.go [hunk 2]
```diff
@@ -121,85 +130,136 @@ func liquidsoapRequest(path string) (message string, err error) {
...
+			resp, err := http.Get("http://" + c.StreamAddress + c.StreamPort + "/status-json.xsl")
+			if resp != nil {
+				defer resp.Body.Close()
+			}
+			if err != nil {
+				clog.Error("icecastMonitor", "Unable to stream data from the Icecast service.", err)
+				icecastDataReset()
+				continue
+			}
 			defer resp.Body.Close()
```
Comment (za419) [inline cadence/server/api_actions.go:195]: Is there a reason we do this twice if `resp != nil`?

Comment (kenellorando) [reply cadence/server/api_actions.go:194]: There is not. I just got rid of "check if `resp != nil` " pattern.

File: cadence/server/api_actions.go (hunk 1 above)
Comment (kenellorando) [reply cadence/server/api_actions.go:150]: Can you give more detail on the latter?

Comment (za419) [reply cadence/server/api_actions.go:150]: Well, it's my understanding that the `fsnotify` watcher only fires off if the directory we're watching gets modified.

That's fine if the filesystem propagates changes (writing a file into a subdirectory fires a modification event on the parent), or if the administrator knows this is a thing and puts files flat in the music dir, but if an enterprising filesystem decides it doesn't need to be nice and make sure the parent directory gets a "change" when a subdirectory gets a new file, then that would mean using it for the music dir would leave gaps in the populator's awareness of change.

Basically - Should we/might we/is it worth switching to something like `rfsnotify` in the future which handles subdirectories explicitly, and/or is my understanding incorrect?

**Review by za419 — APPROVED** [review]

File: cadence/server/api_actions.go (hunk 1 above)
Comment (kenellorando) [reply cadence/server/api_actions.go:150]: I think I understand now. The target directory is not presently recursively searched during population. Only the root of the target of searched. 

Though, of course, I'm now imagining a user who wants to simply hook up their Cadence installation to their `~/Music/` folder, which is populated with hundreds of subfolders of albums. 

Yeah. We should plan to revisit this eventually and implement both recursive population and monitoring.

**Pull request merged**